### PR TITLE
Fix mixed precision issue in TF DistilBert

### LINKED ISSUE #6858: Remove hard-coded uses of float32 to fix mixed precision use in Distilbert
In this commit [Remove hard-coded uses of float32 to fix mixed precision use](https://github.com/huggingface/transformers/commit/4fca874ea995f3d23ad7062b07b5ed7c4f87c0cd#diff-e3ab4f29f29fe1d243a6b55fafaab097), the mixed precision issue is fixed for [modeling_tf_bert.py](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_tf_bert.py).

However, for [modeling_tf_distilbert.py](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_tf_distilbert.py#L171), the line 171 is not fixed yet, and we get 

>    173         embeddings = inputs_embeds + position_embeddings  # (bs, max_seq_length, dim)
>    --> 174         embeddings = self.LayerNorm(embeddings)  # (bs, max_seq_length, dim)
>    InvalidArgumentError: cannot compute AddV2 as input #1(zero-based) was expected to be a bfloat16 tensor but is a float tensor

while using `mixed_bfloat16 mixed precision` with `TPU`.

A very quick fix is the same as the fix for `modeling_tf_bert.py`:

    position_embeddings = tf.cast(self.position_embeddings(position_ids), inputs_embeds.dtype)

@schmidek

### PR DESCRIPTION
**Pull request by ydshieh**
Fix mixed precision issue in TF DistilBert by removing hard-coded uses of float32.

Fixes #6858

### CODE REVIEW

**Comment by xuxingya** [comment]
I run a test with this change on my ubuntu 18.04 with a 2080Ti GPU:
```
from tensorflow.keras.layers import Input, Embedding, Bidirectional, GRU, Dense
from tensorflow.keras.models import Model
from transformers import TFDistilBertModel
from tensorflow.keras.mixed_precision import experimental as mixed_precision
policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)

bert = TFDistilBertModel.from_pretrained('distilbert-base-uncased')
inputs = Input(shape=(None,), dtype='int32')
bert_out = bert(inputs)[0]
output = Dense(9, activation='softmax', dtype='float32')(bert_out)
model = Model(inputs, output)
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
model.summary()
x = [[5, 2, 3] * 3] * 100
y = [[1, 2, 3] * 3] * 100
model.fit(x=x, y=y, epochs=20, batch_size=16)
```
And get error info:
```
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    bert = TFDistilBertModel.from_pretrained('distilbert-base-uncased')
  File "/home/xingya/transformers/src/transformers/modeling_tf_utils.py", line 602, in from_pretrained
    model(model.dummy_inputs, training=False)  # build the network with dummy inputs
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 615, in call
    outputs = self.distilbert(inputs, **kwargs)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 508, in call
    tfmr_output = self.transformer(
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 401, in call
    layer_outputs = layer_module(hidden_state, attn_mask, head_mask[i], output_attentions, training=training)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 355, in call
    ffn_output = self.ffn(sa_output, training=training)  # (bs, seq_length, dim)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 304, in call
    x = self.activation(x)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 968, in __call__
    outputs = self.call(cast_inputs, *args, **kwargs)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py", line 420, in call
    return self.activation(inputs)
  File "/home/xingya/transformers/src/transformers/modeling_tf_distilbert.py", line 79, in gelu
    cdf = 0.5 * (1.0 + tf.math.erf(x / tf.math.sqrt(2.0)))
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 984, in binary_op_wrapper
    return func(x, y, name=name)
  File "/home/xingya/.conda/envs/transformers/lib/python3.8/site-packages/tensorflow/python/ops/math_ops.py", line 1081, in _truediv_python3
    raise TypeError("x and y must have the same dtype, got %r != %r" %
TypeError: x and y must have the same dtype, got tf.float16 != tf.float32
```
I made a modification to L299:
`self.activation = (
            tf.keras.layers.Activation(gelu, dtype='float32') if config.activation == "gelu" else tf.keras.activations.relu
        )`
And then the model began to train, however the loss don't decrease and the accuracy is always 0:
```
7/7 [==============================] - 0s 28ms/step - loss: 2.1972 - accuracy: 0.0000e+00
Epoch 2/20
7/7 [==============================] - 0s 29ms/step - loss: 2.1972 - accuracy: 0.0000e+00
Epoch 3/20
7/7 [==============================] - 0s 30ms/step - loss: 2.1972 - accuracy: 0.0000e+00
Epoch 4/20
7/7 [==============================] - 0s 31ms/step - loss: 2.1972 - accuracy: 0.0000e+00
```

I have trid this code in float32 precision, and it works. 
```
Epoch 1/20
7/7 [==============================] - 0s 31ms/step - loss: 2.5418 - accuracy: 0.2800
Epoch 2/20
7/7 [==============================] - 0s 33ms/step - loss: 1.2452 - accuracy: 0.3356
Epoch 3/20
7/7 [==============================] - 0s 31ms/step - loss: 1.1438 - accuracy: 0.3267
Epoch 4/20
7/7 [==============================] - 0s 33ms/step - loss: 1.1219 - accuracy: 0.3400
```

**Comment by chiapas** [comment]
@xuxingya , the accuracy not improved during training is due to a line 

    > scores = scores - 1e30 * (1.0 - mask)

while `1e30` with `half precision` will cause `nan` values. I am still trying to figure out a way to deal with it.

**Comment by chiapas** [comment]
@xuxingya Would you mind to run the test on your side again, please?

**Comment by xuxingya** [comment]
@chiapas Yes, I run the test and now it's fine.

**Review by patrickvonplaten — APPROVED** [review]
Awesome - very clean fix :-)

File: src/transformers/modeling_tf_distilbert.py [hunk 1]
```diff
@@ -261,9 +263,12 @@ def unshape(x):
         scores = tf.matmul(q, k, transpose_b=True)  # (bs, n_heads, q_length, k_length)
         mask = tf.reshape(mask, mask_reshape)  # (bs, n_heads, qlen, klen)
         # scores.masked_fill_(mask, -float('inf'))            # (bs, n_heads, q_length, k_length)
-        scores = scores - 1e30 * (1.0 - mask)
 
-        weights = tf.nn.softmax(scores, axis=-1)  # (bs, n_heads, qlen, klen)
+        scores_dtype = scores.dtype
+        # calculate `scores` in `tf.float32` to avoid numeric overflow
+        scores = tf.cast(scores, dtype=tf.float32) - 1e30 * (1.0 - tf.cast(mask, dtype=tf.float32))
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_distilbert.py:269]: Yes - softmax should stay in float32!

**Pull request merged**